src/main/kotlin/exercise4/task2/TransactionalBankAccount.kt

```diff
@@ -0,0 +1,225 @@
+ *      Status: [TRANSACTION_STATUS]
+ *
+ *      Transaction Date: [PRETTY_FORMATED_TRANSACTION_DATE]
+ *      Transaction Type: [TRANSACTION_TYPE]
+ *      Amount: [TRANSACTION_AMOUNT]
+ *      Old Balance: [OLD_BALANCE]
+ *      New Balance: [NEW_BALANCE]
+ *      Status: [TRANSACTION_STATUS]
+ *      ```
+ *     If there is no transaction information should be printed as following:
+ *     ```
+ *      Account Holder: [ACCOUNT_HOLDER_NAME]
+ *      Account Number: [ACCOUNT_NUMBER]
+ *      Balance: [ACCOUNT_BALANCE]
+ *
+ *      Transactions:
+ *
+ *      No transactions recorded.
+ *      ```
+ */
+
+private val currentTime: LocalDateTime get() = LocalDateTime.now()
+
+private fun LocalDateTime.prettyPrint(): String {
+    val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS")
+    return this.format(formatter)
+}
+
+fun main() {
+    // Creating a Transactional Bank Account
+    val account = TransactionalBankAccount("123456789", "John Doe")
+
+    // Making some transactions (successful and unsuccessful)
+    account.deposit(1000.0)
+    account.withdraw(500.0)
+    account.deposit(-100.0)
+    account.withdraw(2000.0)
+    account.deposit(500.0)
+    account.withdraw(100.0)
+
+    // Displaying updated account information
+    account.displayAccountInfo()
+
+    println("All Transactions Dates:")
+    account.getAllTransactions().forEach { println(it.getTransactionDate().prettyPrint()) }
+
+    println("\nTransactions with Amount > 600:")
+    account.getAllTransactionsBy { it.getAmount() > 600 }.forEach{ println(it.getAmount()) }
+
+    println("\nTransactions in the Last Hour:")
+    account.getTransactionsBetween(LocalDateTime.now().minusHours(1), LocalDateTime.now()).forEach { println(it.getAmount()) }
+
+    println("\nAll Successful Transactions:")
+    account.getAllSuccessfulTransactions().forEach { println(it.getAmount()) }
+
+    println("\nAll Failed Transactions:")
+    account.getAllFailedTransactions().forEach { println(it.getAmount()) }
+
+    println("\nAll Successful Deposits:")
+    account.getAllSuccessfulDeposits().forEach { println(it.getAmount()) }
+
+    println("\nAll Successful Withdrawals:")
+    account.getAllSuccessfulWithdrawals().forEach { println(it.getAmount()) }
+
+    println("\nAll Failed Deposits:")
+    account.getAllFailedDeposits().forEach { println(it.getAmount()) }
+
+    println("\nAll Failed Withdrawals:")
+    account.getAllFailedWithdrawals().forEach { println(it.getAmount()) }
+}
+
+class TransactionalBankAccount( accountNumber: String, accountHolderName: String, balance: Double = 0.0):
+    BankAccount(accountNumber, accountHolderName, balance) {
+
+    private var transactions = mutableListOf<Transaction>()
+
+    fun getAllTransactions(): List<Transaction> {
+        transactions.sortByDescending { it.getTransactionDate() }
+        return transactions
+    }
+    fun getAllTransactionsBy(predicate: (Transaction) -> Boolean) = transactions.filter(predicate)
```

Comment:
For example, this method does not leak the `transactions` property because the `filter` function returns a new list that contains only filtered elements.